### PR TITLE
[JIT] Always transform `AND(X, CNS(-1))` to `X`

### PR DESCRIPTION
**Pull request by TIHan**
## Description

We should always try to lower `AND(X, CNS(-1))` to `X` if possible. We do this transformation in lowering to pick up this pattern as a result of decomposing longs on 32bit archs.

### CODE REVIEW

**Comment by jakobbotsch** [comment]
Can this optimization be done in lowering instead so that all targets benefit equally? Are there any benefits to the existing x64 peephole for doing the optimization in codegen?

**Comment by tannergooding** [comment]
What's leading to the `x & ~0` in the first place? Is this something specific that's being introduced in lowering or is it a more general missing opt in morph/constant folding?

**Comment by TIHan** [comment]
@tannergooding @jakobbotsch 

Good points, since we don't have to worry about the upper 32 bits, we could do the transform earlier. I have an idea where.

**Comment by jakobbotsch** [comment]
> since we don't have to worry about the upper 32 bits, we could do the transform earlier

Even on 64-bit we can do the optimization for `TYP_INT` `GT_AND` on IR nodes and I'd be curious to see if the peephole isn't just a more conservative version of that.

**Comment by TIHan** [comment]
> Even on 64-bit we can do the optimization

I don't think we can - I remember trying to do (x & -1) optimization in an earlier phase before; ran into issues of the upper 32bits, which is why I did it as a peephole.

**Comment by jakobbotsch** [comment]
> I don't think we can - I remember trying to do (x & -1) optimization in an earlier phase before; ran into issues of the upper 32bits, which is why I did it as a peephole.

I'd be interested to see the example, that certainly sounds odd or like there is a bug somewhere else.

**Comment by TIHan** [comment]
So, we definitely should do the transformation in lowering for x86 because we catch cases of `(x & -1)` after decomposing longs.

**Comment by TIHan** [comment]
@dotnet/jit-contrib This is ready PTAL @BruceForstall

**Comment by BruceForstall** [comment]
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=176379&view=ms.vss-build-web.run-extensions-tab) -- significant asm diffs on x86, a few on x64

File: src/coreclr/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -241,6 +241,54 @@ GenTree* Lowering::LowerMul(GenTreeOp* mul)
...
+
+#ifdef TARGET_64BIT
+    if (!node->TypeIs(TYP_LONG) || !op2->TypeIs(TYP_LONG))
+        return nullptr;
+#else  // TARGET_64BIT
+    assert(op2->TypeIs(TYP_INT));
+
+    if (!node->TypeIs(TYP_INT))
+        return nullptr;
+#endif // !TARGET_64BIT
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerxarch.cpp:276]: I still don't understand why we wouldn't be able to do this for `TYP_INT` on 64-bit. Can you share an example please?

Comment (TIHan) [reply src/coreclr/jit/lowerxarch.cpp]: I will try to enable it again and run Fuzzlyn.

Comment (TIHan) [reply src/coreclr/jit/lowerxarch.cpp]: Thought about it some more, and I think you might be right. Whatever is using the result of `AND(VAR, -1)` shouldn't be caring about the upper 32bits anyway; therefore, it should be fine to just return `VAR`.

File: src/coreclr/jit/lowerxarch.cpp [hunk 2]
```diff
@@ -241,6 +241,54 @@ GenTree* Lowering::LowerMul(GenTreeOp* mul)
...
+//----------------------------------------------------------------------------------------------
+// Lowering::TryLowerAndNegativeOne:
+//    If safe, lowers a tree AND(X, CNS(-1)) to X.
+//
+// Arguments:
+//    node - GT_AND node of integral type
+//
+// Return Value:
+//    Returns the replacement node if one is created else nullptr indicating no replacement
+GenTree* Lowering::TryLowerAndNegativeOne(GenTreeOp* node)
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerxarch.cpp:253]: We can have this on arm32/arm64 too and get benefits there as well.

Comment (TIHan) [reply src/coreclr/jit/lowerxarch.cpp:253]: Yea, I can move this into lowering.cpp.

**Comment by TIHan** [comment]
@dotnet/jit-contrib @jakobbotsch this is ready again. currently superpmi replay failure is unrelated

**Review by jakobbotsch — APPROVED** [review]
LGTM!

**Comment by TIHan** [comment]
Dope

**Pull request merged**